www/blog/2023-01-08-tinyrpc-client.mdx

```diff
@@ -0,0 +1,210 @@
+---
+slug: tinyrpc-client
+title: Writing a tiny tRPC client
+author: Julius
+author_title: tRPC Contributor
+author_url: https://twitter.com/jullerino
+author_image_url: https://github.com/juliusmarminge.png
+---
+
+Ever wondered how tRPC works? Maybe you want to start contributing to the project but is freightened by the internals? The aim of this post is to familiarize you with the internals of tRPC by writing a minimal client that covers the big parts of how tRPC works.
+
+## Overview
+
+Let's assume we have simple tRPC router that looks like this:
+
+```ts
+const appRouter = t.router({
+  post: t.router({
+    byId: t.procedure
+      .input(z.string())
+      .query(({ input }) => posts.find((p) => p.id === input)),
+    create: t.procedure
+      .input(z.object({ title: z.string() }))
+      .mutation(({ input }) => {
+        const post = { id: uuid(), ...input };
+        posts.push(post);
+        return post;
+      }),
+  }),
+});
+```
+
+The goal of our client is to mimic this object structure on our client so that we can call procedures like:
+
+```ts
+const post = await client.post.byId.query('123');
+const newPost = await client.post.create.mutate({ title: 'Foo' });
+```
+
+To do this, tRPC uses [proxies](https://mdn.io/proxy) to augment the object structure with the `.query` and `.mutate` methods on them. Proxies allows us to intercept and modify the operations for the path you are accessing.
```

Comment:
Could be phrased more entertaining I think:

"To do this, tRPC uses a combination of a `Proxy`-object and some TypeScript magic where we actually lie to you about what it does (more on that later)"